enterprise/backend/src/metabase_enterprise/metabot_v3/dummy_tools.clj

```diff
@@ -0,0 +1,75 @@
+(ns metabase-enterprise.metabot-v3.dummy-tools
+  (:require
+   [cheshire.core :as json]
+   [metabase-enterprise.metabot-v3.tools.create-dashboard-subscription]
+   [metabase-enterprise.metabot-v3.tools.query]
+   [metabase-enterprise.metabot-v3.tools.who-is-your-favorite]
+   [metabase.api.common :as api]
+   [toucan2.core :as t2]))
+
+(defn- get-current-user
+  [_ _ context]
+  {:output (if-let [{:keys [id email first_name last_name]}
+                    (or (some-> api/*current-user* deref)
+                        (t2/select-one [:model/User :id :email :first_name :last_name] api/*current-user-id*))]
+             {:id id
+              :name (str first_name " " last_name)
+              :email_address email}
+             {:error "current user not found"})
+   :context context})
+
+(defn- get-dashboard-details
+  [_ {:keys [dashboard_id]} context]
+  {:output (or (t2/select-one [:model/Dashboard :id :description :name] dashboard_id)
+               {:error "dashboard not found"})
+   :context context})
+
+(defn- dummy-tool-messages
+  [tool-id arguments content]
+  (let [call-id (random-uuid)]
+    [{:content    nil
+      :role       :assistant,
+      :tool_calls [{:id        call-id
+                    :name      tool-id
+                    :arguments (json/generate-string arguments)}]}
+
+     {:content      (json/generate-string content)
+      :role         :tool,
+      :tool_call_id call-id}]))
+
+(defn- dummy-get-current-user
+  [context]
+  (let [content (:output (get-current-user :get-current-user {} context))]
+    (dummy-tool-messages :get-current-user {} content)))
+
+(defn- dummy-get-dashboard-details
+  [context]
+  (reduce (fn [messages viewed]
+            (if-let [dashboard-id (when (= (:type viewed) :dashboard)
+                                    (:ref viewed))]
+              (let [arguments {:dashboard_id dashboard-id}
+                    content (-> (get-dashboard-details :get-dashboard-details arguments context)
+                                :output)]
+                (into messages (dummy-tool-messages :get-dashboard-details arguments content)))
+              messages))
+          []
+          (:user-is-viewing context)))
+
+(def ^:private dummy-tool-registry
+  "Get metadata about the available dummy tools. "
+  [dummy-get-current-user
+   dummy-get-dashboard-details])
+
+(defn invoke-dummy-tools
+  "Invoke `tool` with `context` if applicable and return the resulting context."
+  [context]
+  (let [context (or (not-empty context)
+                    ;; hallucinate the user is viewing dashboard with ID 14
+                    {:user-is-viewing [{:type :dashboard
+                                        :ref 14
+                                        :parameters []
+                                        :is-embedded false}]})]
```

Comment:
This should provided by the FE and it should be a semantically meaningful description of the state of the FE (the users context). The format should be generic enough that doesn't change often, it's the task of the BE to turn it into input useful for the  AI service. Currently, the BE generates dummy tool calls based on it.

Reply:
super micro but we should use a word other than `hallucinate` here since since it has a very specific meaning in LLM land that is different.

Reply:
**question (not-blocking)**:
What was the reason that we decided to construct the dummy tool message history from the BE rather than just sending those in the message history from the frontend? 
It is something very frontend specific and moving forward, having it in the BE would require both a FE and a BE change if we needed more dummy tool calls + results in the conversation history.

Right now, the `context` is being sent from the FE and used to construct those. We also forward the context to the LLM. If we use the context both for FE<->BE communication AND for the LLM, we need to pre-filter it before we forward it to the LLM. Or we have a separate  LLM `context` constructed in the BE. We should just avoid sending over context that we did not test with (even if it is "just" a dashboard id :)

Reply:
The idea is that the FE provides the state (and possibly historical) information in a generic format that doesn't change often. If a new concept is introduced in the FE, it will also be described in this generic format and it will be ignored until we enable its processing in the BE. This is intentional: we want to prevent misleading the agent and the job of the BE is to translated the generic description from the FE to something understandable for the LLM. (As you said, we don't want to send untested things to the LLM.)

We are not using the history directly, because we hope we can convey the necessary information as dummy tool calls, and at some point these dummy tools might become real tools available for the agent to use.

Reply:
> We are not using the history directly, because we hope we can convey the necessary information as dummy tool calls, and at some point these dummy tools might become real tools available for the agent to use.

My mind model around the dummy tool calls was that we simulate a conversation state to shortcut the agent work. This could include dummy tools that don't really exist **but also calls of real tools, e.g.:**

**User is on a home page (that has not set up a dashboard).**
* The conversation history would purely contain a `get-current-user` tool call (since we shortcut that for the agent that it does not need to call that first)
* If the user asks for the value of a metric, the agent would 
  * search for it (tool)
  * query it (tool)
  * respond back with the results
  
 **Scenario 2: User is on a metric page**
 * The conversation history would already contain a tool calls for the "search" of the metric -> we shortcut and tell the agent: You already searched for that metric, the user is likely referring to that. (This is the simulation part)
 * If the user asks for the value of the metric, the agent would
   * query it (tool)
   * respond back with the results
   
 This is independent of whether those tools really exist already or if they are pure dummy implementations. The `get-current-metric` one for example will likely be replaced by the metric selection tool I am working on now (the one we talked about yesterady).
 
That is why I was thinking in those cases the FE should just inject those into the conversation history. 
Does this change the way we think about this?

Reply:
We had a discussion and came to the conclusion that the current way of handling dummy tool calls is OK.
1. Their result doesn't have to show up in the history like real tool invocations do.
1. We are OK with having to process the info from the FE in the BE first: this extra steps makes it easier to control what information actually reaches the LLM.